Add a sidebar filter for organisations, taxons and search

theleebriggs opened the pull request:
### Motivation
As part of ongoing design iteration we are breaking up the apps main view into a few areas - one of these is a sidebar containing filtering tools e.g. search box, organisation list. This PR is to refactor currently separate areas of the app e.g. standalone filter page, search box - into one cohesive component which users will find easier to use and navigate the app content with.

https://trello.com/c/spNc2fUv/217-5-revise-alpha-presentation-add-the-filter-area

### Description
This work adds a new `sidebar` partial containing a form with fields for filtering by `title`, `organisation` and `taxonomy`. It also consolidates the feature testing for searching/filtering.

As part of the cleanup it removes all previous `content_items#filter` work and sets the apps root to be `content_items#index` in a move to give users a speedier gateway to the content they may be interested in.

### Before
![before](https://cloud.githubusercontent.com/assets/6338228/24406843/283dac94-13c1-11e7-9caf-b5e44c7ed578.png)

### After
![after-2](https://cloud.githubusercontent.com/assets/6338228/24415612/ed8e644e-13d9-11e7-8376-5c6799ea142f.png)

pmanrubia commented on spec/views/application/_sidebar.html.erb_spec.rb:
```diff
@@ -0,0 +1,42 @@
+      assign(:organisations, organisations)
+      render
+
+      expect(rendered).to have_selector('form select#organisation_id option', count: 3)
```
What about checking here that we are using the organisation title to render the combo?

theleebriggs replied:
Addressed now.

pmanrubia commented on app/controllers/content_items_controller.rb:
```diff
@@ -24,6 +25,11 @@ def filter
 
 private
 
+  def populate_filters
```
I would not include the word `filters` in the controller. `filters` are view related

What about splitting up the method into:

```ruby
before_action :set_all_organisations, only: :index
before_action :set_all_taxonomies, only: :index

theleebriggs replied:
Addressed now.

pmanrubia commented on spec/controllers/content_items_controller_spec.rb:
```diff
@@ -41,6 +41,16 @@
+
+    it "assigns the lists of taxonomies and organisations ordered ASC by title" do
+      organisations = [{ title: "ABC" }, { title: "BCD" }].map { |o| create(:organisation, o) }
+      taxons = [{ title: "ABC" }, { title: "BCD" }].map { |t| create(:taxonomy, t) }
```
The title is not being used in the expectation, so what about?
```ruby
organisations = create_list(:organisations, 1)
taxonomy = create_list(:taxonomy, 1)
```

theleebriggs replied:
The title is used to ensure the list returned is in alphabetical order - but I could make the test clearer by creating them in reverse order then testing the title of the first one i.e. it should be `"ABC"`?

theleebriggs replied:
Made test explicit in testing number or orgs/taxons PLUS that they were ordered alphabetically.

pmanrubia replied:
👍

pmanrubia commented on spec/features/filter_spec.rb:
```diff
@@ -0,0 +1,84 @@
+      visit "/content_items"
+      fill_in 'query', with: 'some text'
+
+      click_on "Filter"
```
Very minor: Could you please check this file and move `click_on "Filter"` 2 lines above the expectation, and next to the user interactions

pmanrubia commented on spec/views/content_items/index.html.erb_spec.rb:
```diff
@@ -50,109 +51,77 @@
+      it "renders item last updated" do
         Timecop.freeze(Time.parse('2016-3-20')) do
-          content_items[0].public_updated_at = Time.parse('2016-1-20')
+          allow(content_items[0]).to receive(:public_updated_at).and_return(Time.parse('2016-1-20'))
```
Any reason for mocking instead of assigning the property with a setter?

theleebriggs replied:
This is done partly for consistency with the other assertions in this file - but also for robustness if this value switches to a decorator or similar - the view should only ensure it rendered whatever is returned from that `:message` being sent.

pmanrubia replied:
👍

theleebriggs commented on spec/views/application/_sidebar.html.erb_spec.rb:
```diff
@@ -0,0 +1,42 @@
+  end
+
+  context "filtering organisations" do
+    it "will render select boxes with options for 2 organisations and 1 empty if supplied with list of 2 organisations" do
```
Addressed this now.

theleebriggs commented on spec/views/application/_sidebar.html.erb_spec.rb:
```diff
@@ -0,0 +1,42 @@
+  end
+
+  context "filtering by taxonomies" do
+    it "will render select boxes with with options for 2 taxonomies and 1 empty if supplied with list of 2 taxonomies" do
```
Addressed now.

theleebriggs commented:
@pmanrubia I've addressed your comments now - Org column has been removed and heading for "Page Views" updated in both index and show. I have kept it as "Page views (1 month)" rather than "(1m)" as it could be confused with "(1 million)".

pmanrubia commented:
LGTM. Good work 👍 
Could you please update the screen capture from the PR?

theleebriggs commented:
Screenshot updated - merging now. Thanks for the speedy review 💯
